File: chronicle-pc/frontend/src/brain/capGraphNodes.ts
import type { BrainGraph, GraphEdge, GraphNode } from '../api/types'

export const DEFAULT_NODE_CAP = 500

export interface CappedGraph {
  nodes: GraphNode[]
  edges: GraphEdge[]
  totalNodes: number
  capped: boolean
}

/** Score by degree (primary) then recency (ts), keep pinned, cap rendered nodes. */
export function capGraphNodes(
  graph: BrainGraph,
  limit: number = DEFAULT_NODE_CAP,
): CappedGraph {
  const all = (graph.nodes || []).filter((n) => !n.hidden)
  const totalNodes = all.length
  if (totalNodes <= limit) {
    const ids = new Set(all.map((n) => n.id))
    const edges = (graph.edges || []).filter((e) => ids.has(e.from) && ids.has(e.to))
    return { nodes: all, edges, totalNodes, capped: false }
  }

  const degree = new Map<string, number>()
  for (const e of graph.edges || []) {
    degree.set(e.from, (degree.get(e.from) || 0) + 1)
    degree.set(e.to, (degree.get(e.to) || 0) + 1)
  }

  const score = (n: GraphNode): number => {
    const deg = degree.get(n.id) || 0
    const ts = n.ts ? Date.parse(n.ts) : 0
    const recency = Number.isFinite(ts) ? ts / 1e13 : 0
    return deg * 1000 + recency + (n.pinned ? 1e6 : 0) + (n.weight ?? 0)
  }

  const pinned = all.filter((n) => n.pinned)
  const rest = all
    .filter((n) => !n.pinned)
    .sort((a, b) => score(b) - score(a))
  const kept = [...pinned, ...rest].slice(0, limit)
  const ids = new Set(kept.map((n) => n.id))
  const edges = (graph.edges || []).filter((e) => ids.has(e.from) && ids.has(e.to))
  return { nodes: kept, edges, totalNodes, capped: true }
}

File: chronicle-android/app/src/main/java/com/chronicle/app/brain/CapGraphNodes.kt
package com.chronicle.app.brain

/** Default max nodes rendered on the mind map (Mac ForceGraph parity). */
const val DEFAULT_NODE_CAP = 500

data class CappedGraph(
    val nodes: List<GraphNode>,
    val edges: List<GraphEdge>,
    val totalNodes: Int,
    val capped: Boolean,
)

/**
 * Score by degree (primary) then recency (ts), keep pinned, cap rendered nodes.
 * Mirrors chronicle-pc/frontend/src/brain/capGraphNodes.ts.
 */
fun capGraphNodes(
    graph: BrainGraph,
    limit: Int = DEFAULT_NODE_CAP,
): CappedGraph {
    val all = graph.nodes.filter { !it.hidden }
    val totalNodes = all.size
    if (totalNodes <= limit) {
        val ids = all.map { it.id }.toSet()
        val edges = graph.edges.filter { it.from in ids && it.to in ids }
        return CappedGraph(all, edges, totalNodes, capped = false)
    }

    val degree = mutableMapOf<String, Int>()
    for (e in graph.edges) {
        degree[e.from] = (degree[e.from] ?: 0) + 1
        degree[e.to] = (degree[e.to] ?: 0) + 1
    }

    fun score(n: GraphNode): Double {
        val deg = degree[n.id] ?: 0
        val ts = n.ts?.let { runCatching { java.time.Instant.parse(it).toEpochMilli() }.getOrNull() } ?: 0L
        val recency = if (ts > 0) ts / 1e13 else 0.0
        return deg * 1000.0 + recency + (if (n.pinned) 1e6 else 0.0) + n.weight
    }

    val pinned = all.filter { it.pinned }
    val rest = all.filter { !it.pinned }.sortedByDescending { score(it) }
    val kept = (pinned + rest).take(limit)
    val ids = kept.map { it.id }.toSet()
    val edges = graph.edges.filter { it.from in ids && it.to in ids }
    return CappedGraph(kept, edges, totalNodes, capped = true)
}
